syntax = "proto3";

package aggregation_server;

enum Status {
  STATUS_UNSPECIFIED = 0;
  OK = 1;
  ALREADY_CONFIGURED = 2;
  NOT_FOUND = 3;
  MIN_WORKERS = 4;
  TIMEOUT = 5;
  ERROR = 6;
}

service AggregationServer {
  rpc Configure (ConfigureRequest) returns (ConfigureResponse);
  rpc Unregister (UnregisterRequest) returns (UnregisterResponse);
  rpc Aggregate (AggregateRequest) returns (AggregateResponse);
  rpc Cleanup (CleanupRequest) returns (CleanupResponse);
}

message ConfigureRequest {
  string request_id = 1;
  int32 num_of_workers = 2;
}

message ConfigureResponse {
  Status status = 1;
  string status_message = 2;
}

message Operation {
  string aggregation_type = 1;
  repeated double vectors = 2;
  bytes tensor = 3;
}

message AggregateRequest {
  string request_id = 1;
  int32 step = 2;
  repeated Operation operations = 3;
}

message AggregateResponse {
  repeated double results = 1;
  repeated int32 offsets = 2;
  repeated bytes tensors = 3;
}

message CleanupRequest {
  string request_id = 1;
}

message CleanupResponse {
  Status status = 1;
  string status_message = 2;
}

message UnregisterRequest {
  string request_id = 1;
}

message UnregisterResponse {
  Status status = 1;
  string status_message = 2;
  int32 remaining_workers = 3;
}
